src/mango.ts

```diff
@@ -0,0 +1,493 @@
+      const fillQty = (hedgeSide === HedgeSide.buy ? 1 : -1) * quantity;
+      hedgeDeltaTotal += fillQty;
+      console.log(`${scalper.symbol} Delta Filled ${hedgeSide} ${hedgeProduct} Qty ${fillQty} \
+          Price ${price} Remaining ${hedgeDeltaTotal} ID ${deltaOrderId} ${new Date().toUTCString()}`);
+    }
+  };
+
+  // Setup a listener for the order.
+  if (fillFeed.readyState === WebSocket.OPEN) {
+    fillFeed.addEventListener('message', deltaFillListener);
+    console.log(scalper.symbol, 'Listening For Delta Hedges');
+  } else {
+    console.log(scalper.symbol, 'Websocket State', fillFeed.readyState);
+  }
+
+  console.log(`${scalper.symbol} ${hedgeSide} ${hedgeProduct} ${Math.abs(hedgeDeltaClip)} \
+    Limit: ${hedgePrice} # ${deltaHedgeCount} ID ${deltaOrderId}`);
+
+  try {
+    if (hedgeProduct === HedgeProduct.Spot) {
+      const spotHedgeSide = hedgeSide === HedgeSide.buy ? Serum3Side.bid : Serum3Side.ask;
+      await mangoClient.serum3PlaceOrder(
+        mangoGroup,
+        mangoAccount,
+        new PublicKey(OPENBOOK_MKT_MAP.get(scalper.symbol)),
+        spotHedgeSide,
+        hedgePrice,
+        Math.abs(hedgeDeltaClip),
+        Serum3SelfTradeBehavior.decrementTake,
+        Serum3OrderType.limit,
+        deltaOrderId,
+        10,
+      );
+    } else {
+      const perpHedgeSide = hedgeSide === HedgeSide.buy ? PerpOrderSide.bid : PerpOrderSide.ask;
+      await mangoClient.perpPlaceOrder(
+        mangoGroup,
+        mangoAccount,
+        perpMarket.perpMarketIndex,
+        perpHedgeSide,
+        hedgePrice,
+        Math.abs(hedgeDeltaClip),
+        undefined,
+        deltaOrderId,
+        PerpOrderType.limit,
+      );
+    }
+  } catch (err) {
+    console.log(scalper.symbol, 'Failed to place order', err, err.stack);
+  }
+
+  // Wait the twapInterval of time to see if the position gets to neutral.
+  console.log(scalper.symbol, 'Scan Delta Fills for ~', TWAP_INTERVAL_SEC, 'seconds');
+  await waitForFill(
+    (_) => Math.abs(hedgeDeltaTotal) < (deltaThreshold),
+    TWAP_INTERVAL_SEC,
+  );
+
+  // Cleanup listener.
+  fillFeed.removeEventListener('message', deltaFillListener);
+
+  if (Math.abs(hedgeDeltaTotal) < (deltaThreshold)) {
+    fillFeed.removeEventListener('message', deltaFillListener);
+    console.log(scalper.symbol, 'Delta Hedge Complete', hedgeDeltaTotal, '<', deltaThreshold);
+    return;
+  }
+
+  // Recursive call, happens when we are not getting fills to get to neutral.
+  await deltaHedgeMango(
+    dipProduct,
+    scalper,
+    mangoClient,
+    mangoAccount,
+    mangoGroup,
+    perpMarket,
+    spotMarket,
+    fillFeed,
+    deltaHedgeCount + 1,
+  );
+}
+
+export async function gammaScalpMango(
+  dipProduct: DIPDeposit[],
+  scalper: Scalper,
+  mangoClient: MangoClient,
+  mangoAccount: MangoAccount,
+  mangoGroup: Group,
+  perpMarket: PerpMarket,
+  fillFeed: WebSocket,
+  gammaScalpCount: number,
+): Promise<void> {
+  await mangoAccount.reload(mangoClient);
```

Comment:
consider a function for these 3 lines called resetMangoState.
Your call. 3 lines repeated is not that much, would only do it so that it is obvious to the reader what is being accomplished at the start

Reply:
Comment added instead